mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,307 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "splitmultipletapes"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Types.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+
+#include <memory>
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_SPLITMULTIPLETAPESPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct SplitMultipleTapesPass : public impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass> {
+    using impl::SplitMultipleTapesPassBase<SplitMultipleTapesPass>::SplitMultipleTapesPassBase;
+
+    unsigned int countTapes(const func::FuncOp &func)
+    {
+        // Count the number of quantum.device operations in a function
+        unsigned int count = 0;
+        func->walk([&](Operation *op) {
+            if (op->getName().getStringRef() == "quantum.device") {
+                count++;
+            }
+        });
+        return count;
+    } // countTapes()
+
+    void collectOperationsForEachTape(
+        const func::FuncOp &func,
+        SmallVector<std::shared_ptr<SmallVector<Operation *>>> &OpsEachTape)
+    {
+        // During tracing, each tape starts with a qdevice_p primitive
+        // This means each tape starts with a quantum.device
+        // and ends with a quantum.device_release
+
+        // The structure of the FuncOp looks like the following:
+        // func.func @circuit(...) -> ... {...} {
+        // preprocessing
+        // quantum.device[...]
+        // tape 1
+        // quantum.device_release
+        // quantum.device[...]
+        // tape 2
+        // quantum.device_release
+        // ... more tapes
+        // post processing and return
+        // }
+
+        // This function parses the operations in the funcop and puts them into
+        // the container as
+        // OpsEachTape = [[preprocessing ops], [tape1 ops], [tape2 ops], ..., [postprocessing ops]]
+
+        for (size_t i = 0; i < OpsEachTape.size(); i++) {
+            OpsEachTape[i] = std::make_shared<SmallVector<Operation *>>();
```

Comment:
I'm curious why did you wrap the inner vector in a pointer?

Reply:
You are right, it doesn't need to be a pointer. It can just be the object...

This reminds me of [Stroustrup's wise words of "Ironically <the problem of not properly deleting, which smart pointers are supposed to solve> can be solved by just not using pointers at all"](https://agorism.dev/book/c%2B%2B/%28C%2B%2B%20In-Depth%20Series%29%20Bjarne%20Stroustrup%20-%20A%20Tour%20of%20C%2B%2B-Addison-Wesley%20Professional%20%282018%29.pdf) (P178)

Reply:
These are now no longer pointers.

Reply:
Nice quote 😄